x-pack/plugins/file_upload/server/capabilities.test.ts

```diff
@@ -0,0 +1,261 @@
+                "show": true,
+              },
+              "management": Object {},
+              "navLinks": Object {},
+            }
+          `);
+  });
+
+  it('registers a capabilities switcher that returns unaltered capabilities when default capabilities are requested', async () => {
+    const coreSetup = coreMock.createSetup();
+    const security = securityMock.createStart();
+    security.authz.mode.useRbacForRequest.mockReturnValue(true);
+    coreSetup.getStartServices.mockResolvedValue([
+      (undefined as unknown) as CoreStart,
+      { security },
+      undefined,
+    ]);
+    setupCapabilities(coreSetup);
+
+    expect(coreSetup.capabilities.registerSwitcher).toHaveBeenCalledTimes(1);
+    const [switcher] = coreSetup.capabilities.registerSwitcher.mock.calls[0];
+
+    const capabilities = {
+      navLinks: {},
+      management: {},
+      catalogue: {},
+      fileUpload: {
+        show: true,
+      },
+    } as Capabilities;
+
+    const request = httpServerMock.createKibanaRequest();
+
+    await expect(switcher(request, capabilities, true)).resolves.toMatchInlineSnapshot(`
+            Object {
+              "catalogue": Object {},
+              "fileUpload": Object {
+                "show": true,
+              },
+              "management": Object {},
+              "navLinks": Object {},
+            }
+          `);
+
+    expect(security.authz.mode.useRbacForRequest).not.toHaveBeenCalled();
+    expect(security.authz.checkPrivilegesDynamicallyWithRequest).not.toHaveBeenCalled();
+  });
+
+  it('registers a capabilities switcher that disables capabilities for underprivileged users', async () => {
+    const coreSetup = coreMock.createSetup();
+    const security = securityMock.createStart();
+    security.authz.mode.useRbacForRequest.mockReturnValue(true);
+
+    const mockCheckPrivileges = jest.fn().mockResolvedValue({ hasAllRequested: false });
+    security.authz.checkPrivilegesDynamicallyWithRequest.mockReturnValue(mockCheckPrivileges);
+    coreSetup.getStartServices.mockResolvedValue([
+      (undefined as unknown) as CoreStart,
+      { security },
+      undefined,
+    ]);
+    setupCapabilities(coreSetup);
+
+    expect(coreSetup.capabilities.registerSwitcher).toHaveBeenCalledTimes(1);
+    const [switcher] = coreSetup.capabilities.registerSwitcher.mock.calls[0];
+
+    const capabilities = {
+      navLinks: {},
+      management: {},
+      catalogue: {},
+      fileUpload: {
+        show: true,
+      },
+    } as Capabilities;
+
+    const request = httpServerMock.createKibanaRequest();
+
+    await expect(switcher(request, capabilities, false)).resolves.toMatchInlineSnapshot(`
+            Object {
+              "fileUpload": Object {
+                "show": false,
+              },
+            }
+          `);
+
+    expect(security.authz.mode.useRbacForRequest).toHaveBeenCalledTimes(1);
+    expect(security.authz.checkPrivilegesDynamicallyWithRequest).toHaveBeenCalledTimes(1);
```

Comment:
__optional nit:__ maybe we can check that this method was indeed invoked with the correct `request` instance? Not a big deal though.